src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -178,83 +220,158 @@ public static IEnumerable<T> EnumerateAllComponentsInScenesAndPrefabs<T>()
         where TAsset : UnityEngine.Object
     {
-        [SerializeField]
-        private TAssetEntry[] assets = null;
+        const string assetContentExtension = ".asset";
 
-        private Dictionary<AssetId, TAssetEntry> lookupByAssetId;
-        private Dictionary<TAsset, TAssetEntry> lookupByAsset;
+#pragma warning disable 414
+        [SerializeField]
+        private NameEntry[] assets = null;
+#pragma warning restore 414
 
-        protected IDictionary<AssetId, TAssetEntry> LookupByAssetId
+        private Dictionary<string, List<AssetId>> LookupByName
         {
             get
             {
-                if (lookupByAssetId == null)
+                if (lookupByName == null)
                 {
-                    if (assets == null)
-                    {
-                        lookupByAssetId = new Dictionary<AssetId, TAssetEntry>();
-                    }
-                    else
+                    lookupByName = new Dictionary<string, List<AssetId>>();
+                    if (assets != null)
                     {
-                        lookupByAssetId = assets.Where(a => a.Asset != null).ToDictionary(a => a.AssetId);
+                        for (int i = 0; i < assets.Length; i++)
+                        {
+                            lookupByName[assets[i].Name] = assets[i].Ids.ToList();
+                        }
                     }
                 }
-                return lookupByAssetId;
-            }
-        }
 
-        protected IDictionary<TAsset, TAssetEntry> LookupByAsset
-        {
-            get
-            {
-                if (lookupByAsset == null)
-                {
-                    if (assets == null)
-                    {
-                        lookupByAsset = new Dictionary<TAsset, TAssetEntry>();
-                    }
-                    else
-                    {
-                        lookupByAsset = assets.Where(a => a.Asset != null).ToDictionary(a => a.Asset);
-                    }
-                }
-                return lookupByAsset;
+                return lookupByName;
             }
         }
 
+#pragma warning disable 414
+        private Dictionary<string, List<AssetId>> lookupByName;
+        private Dictionary<AssetId, AssetCacheEntry> lookupByAssetId = new Dictionary<AssetId, AssetCacheEntry>();
+        private Dictionary<TAsset, AssetCacheEntry> lookupByAsset = new Dictionary<TAsset, AssetCacheEntry>();
+#pragma warning restore 414
+
         public TAsset GetAsset(AssetId assetId)
         {
             using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "GetAsset"))
             {
-                TAssetEntry assetEntry;
-                if (LookupByAssetId.TryGetValue(assetId, out assetEntry))
+                if (assetId == null ||
+                assetId == AssetId.Empty ||
+                assetId.Name == string.Empty)
+                {
+                    return default(TAsset);

```

Comment:
>return [](start = 20, length = 6)

Curious - is it interesting to increment the event duration if the passed in AssetId is invalid? Or is it separately interesting to log that?

Otherwise I assume you can end up getting some zero-duration (or near-zero duration) events which are probably interesting to know separately (i.e. something is "wrong" because an invalid parameter was passed, rather than just because the duration was effectively zero)

Reply:
for now that measure is primarily to understand how much time in total is spent in this function. We could have different event logging for found vs not found, but I don't think that is needed for now.

---
In reply to: [355039776](https://github.com/microsoft/MixedReality-SpectatorView/pull/295#discussion_r355039776) [](ancestors = 355039776)